File: java/com/joshmanisdabomb/lcc/block/FunctionalGrassBlock.java
package com.joshmanisdabomb.lcc.block;

import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.tags.FluidTags;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.Direction;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockReader;
import net.minecraft.world.IWorldReader;
import net.minecraft.world.World;
import net.minecraft.world.lighting.LightEngine;
import net.minecraftforge.common.IPlantable;
import net.minecraftforge.common.PlantType;

import java.util.Random;
import java.util.function.BiFunction;

public class FunctionalGrassBlock extends GrassBlock {

    private final BiFunction<BlockState, Boolean, BlockState> change;

    public FunctionalGrassBlock(BiFunction<BlockState, Boolean, BlockState> change, Properties properties) {
        super(properties);
        this.change = change;
}

    public boolean survives(BlockState state, IWorldReader world, BlockPos pos, boolean water) {
        BlockPos pos2 = pos.up();
        BlockState state2 = world.getBlockState(pos2);
        if (water && world.getFluidState(pos2).isTagged(FluidTags.WATER)) {
            return false;
        } else if (state2.getBlock() == Blocks.SNOW && state2.get(SnowBlock.LAYERS) == 1) {
            return true;
        } else {
            int i = LightEngine.func_215613_a(world, state, pos, state2, pos2, Direction.UP, state2.getOpacity(world, pos2));
            return i < world.getMaxLightLevel();
        }
    }

    @Override
    public void tick(BlockState state, World world, BlockPos pos, Random random) {
        if (!world.isRemote) {
            if (!world.isAreaLoaded(pos, 3)) return;
            if (!survives(state, world, pos, false)) {
                world.setBlockState(pos, this.change.apply(null, null));
            } else {
                if (world.getLight(pos.up()) >= 9) {
                    for(int i = 0; i < 4; ++i) {
                        BlockPos pos2 = pos.add(random.nextInt(3) - 1, random.nextInt(5) - 3, random.nextInt(3) - 1);
                        if (survives(state, world, pos2, true)) {
                            BlockState param = world.getBlockState(pos2);
                            if (param.getBlock() != this) {
                                BlockState set = this.change.apply(param, world.getBlockState(pos2.up()).getBlock() == Blocks.SNOW);
                                if (set != null && set != param) {
                                    world.setBlockState(pos2, set);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    @Override
    public BlockRenderLayer getRenderLayer() {
        return BlockRenderLayer.SOLID;
    }

    @Override
    public boolean canSustainPlant(BlockState state, IBlockReader world, BlockPos pos, Direction side, IPlantable plantable) {
        net.minecraftforge.common.PlantType type = plantable.getPlantType(world, pos.offset(side));

        if (plantable instanceof BushBlock || type == PlantType.Plains) return true;
        return type == PlantType.Beach && (world.getBlockState(pos.east()).getMaterial() == Material.WATER || world.getBlockState(pos.west()).getMaterial() == Material.WATER || world.getBlockState(pos.north()).getMaterial() == Material.WATER || world.getBlockState(pos.south()).getMaterial() == Material.WATER);
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/block/FunctionalGrassBlock.kt
package com.joshmanisdabomb.lcc.block

import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.block.SnowBlock
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.World
import net.minecraft.world.WorldAccess
import net.minecraft.world.chunk.light.ChunkLightProvider
import java.util.*

open class FunctionalGrassBlock(val change: (world: World, source: BlockState?, destination: BlockState, posFrom: BlockPos, posTo: BlockPos, it: FunctionalGrassBlock) -> BlockState?, settings: Settings, val lightRequired: Int = 9) : Block(settings) {

    open fun survives(state: BlockState, world: WorldAccess, pos: BlockPos, water: Boolean): Boolean {
        val pos2 = pos.up()
        val state2 = world.getBlockState(pos2)
        return if (state2.block == Blocks.SNOW && state2[SnowBlock.LAYERS] == 1) {
            true
        } else if (water && state2.fluidState.level == 8) {
            false
        } else {
            val i = ChunkLightProvider.getRealisticOpacity(world, state, pos, state2, pos2, Direction.UP, state2.getOpacity(world, pos2))
            return i < world.maxLightLevel
        }
    }

    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        if (!survives(state, world, pos, false)) {
            world.setBlockState(pos, change(world, null, state, pos, pos, this))
        } else {
            if (world.getLightLevel(pos.up()) >= lightRequired) {
                for (i in 0..3) {
                    val pos2 = pos.method_34592(random.nextInt(3) - 1, random.nextInt(5) - 3, random.nextInt(3) - 1)
                    if (survives(state, world, pos2, true)) {
                        val param = world.getBlockState(pos2)
                        if (param.block != this) {
                            val set = this.change(world, state, param, pos, pos2, this)
                            if (set != null && set != param) {
                                world.setBlockState(pos2, set)
                            }
                        }
                    }
                }
            }
        }
    }

    companion object {

        fun defaultFunction(grass: BlockState, dirt: BlockState) = { world: World, source: BlockState?, destination: BlockState, from: BlockPos, to: BlockPos, it: Block ->
            when {
                source == null -> dirt
                destination.isOf(Blocks.DIRT) -> grass
                else -> null
            }
        }

        fun defaultFunction(dirt: BlockState) = { world: World, source: BlockState?, destination: BlockState, from: BlockPos, to: BlockPos, it: Block ->
            when {
                source == null -> dirt
                destination.isOf(Blocks.DIRT) -> it.defaultState
                else -> null
            }
        }

    }
}